### PR TITLE
chore: improves tests for hot reload of encryptionconfig

### PR DESCRIPTION
**Pull request by nilekhc**
Signed-off-by: Nilekh Chaudhari <1626598+nilekhc@users.noreply.github.com>

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
none
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go [hunk 1]
```diff
@@ -176,41 +176,65 @@ func TestEncryptionProviderConfigCorrect(t *testing.T) {
 	// Creates compound/prefix transformers with different ordering of available transformers.
 	// Transforms data using one of them, and tries to untransform using the others.
 	// Repeats this for all possible combinations.
+	expectedKMSCloseGracePeriod := 46 * time.Second
```
Comment (aramase) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:179]: I think a comment here on the math behind the `expectedKMSCloseGracePeriod = 46s` would be good for anyone who reads the tests

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go [hunk 2]
```diff
@@ -176,41 +176,65 @@ func TestEncryptionProviderConfigCorrect(t *testing.T) {
...
 	correctConfigWithIdentityFirst := "testdata/valid-configs/identity-first.yaml"
 	identityFirstEncryptionConfiguration, err := LoadEncryptionConfig(correctConfigWithIdentityFirst, false, ctx.Done())
 	if err != nil {
 		t.Fatalf("error while parsing configuration file: %s.\nThe file was:\n%s", err, correctConfigWithIdentityFirst)
 	}
+	if identityFirstEncryptionConfiguration.KMSCloseGracePeriod != expectedKMSCloseGracePeriod {
+		t.Fatalf("KMSCloseGracePeriod mismatch (-want +got):\n%s", cmp.Diff(expectedKMSCloseGracePeriod, identityFirstEncryptionConfiguration.KMSCloseGracePeriod))
+	}
 
+	expectedKMSCloseGracePeriod = 32 * time.Second
```
Comment (aramase) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:189]: ditto (add a comment on the math)

applies in all places where `expectedKMSCloseGracePeriod` is set

**Comment by liggitt** [comment]
assigning to @enj for primary review

**Comment by nilekhc** [comment]
Created https://github.com/kubernetes/kubernetes/issues/114999 to track the symlink swap test

File: test/integration/controlplane/transformation/kms_transformation_test.go [hunk 3]
```diff
@@ -569,11 +569,13 @@ resources:
 			if tc.fileUpdateMethod == "symLink" {
 				test, err = newTransformTest(t, encryptionConfig, true, "", true)
 				if err != nil {
+					test.cleanUp()
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply test/integration/controlplane/transformation/kms_transformation_test.go]: removed code

File: test/integration/controlplane/transformation/transformation_test.go [hunk 4]
```diff
@@ -94,24 +93,24 @@ func newTransformTest(l kubeapiservertesting.Logger, transformerConfigYAML strin
 	// create config dir with provided config yaml
 	if transformerConfigYAML != "" && configDir == "" {
 		if e.configDir, e.configParentDir, err = e.createEncryptionConfig(ecSymLink); err != nil {
-			return nil, fmt.Errorf("error while creating KubeAPIServer encryption config: %v", err)
+			return &e, fmt.Errorf("error while creating KubeAPIServer encryption config: %v", err)
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply test/integration/controlplane/transformation/transformation_test.go]: reverted

File: test/integration/controlplane/transformation/transformation_test.go [hunk 5]
```diff
@@ -254,43 +256,50 @@ func (e *transformTest) getRawSecretFromETCD() ([]byte, error) {
...
+		if err = os.WriteFile(symLinkPathForEncryptionConfig, []byte(e.transformerConfig), 0644); err != nil {
+			return tempDir, symLinkTempDir, fmt.Errorf("failed to write encryption config file: %v", err)
 		}
 
-		// create symlink
-		if err := os.Symlink(filepath.Join(parentTempDir, encryptionConfigFileName), filepath.Join(tempDir, encryptionConfigFileName)); err != nil {
-			return tempDir, parentTempDir, fmt.Errorf("failed to create symlink: %v", err)
+		// create a symlink from temp to parent
+		if err = os.Symlink(symLinkTempDir, tempDir); err != nil {
+			return tempDir, symLinkTempDir, fmt.Errorf("failed to create symlink: %w", err)
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply test/integration/controlplane/transformation/transformation_test.go]: removed code

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go (hunk 1 above)
Comment (nilekhc) [reply staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:180]: done

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go (hunk 2 above)
Comment (nilekhc) [reply staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:191]: done

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go [hunk 6]
```diff
@@ -176,41 +176,71 @@ func TestEncryptionProviderConfigCorrect(t *testing.T) {
 	// Creates compound/prefix transformers with different ordering of available transformers.
 	// Transforms data using one of them, and tries to untransform using the others.
 	// Repeats this for all possible combinations.
+	// Math for GracePeriod is explained at - https://github.com/kubernetes/kubernetes/blob/c9ed04762f94a319d7b1fb718dc345491a32bea6/staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go#L140-L143
```
Comment (aramase) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:179]: Isn't this snippet https://github.com/kubernetes/kubernetes/blob/c9ed04762f94a319d7b1fb718dc345491a32bea6/staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go#L159-L163 a better explanation for how the grace period is calculated?

File: test/integration/controlplane/transformation/kms_transformation_test.go [hunk 7]
```diff
@@ -515,35 +516,81 @@ resources:
...
-	test, err = newTransformTest(t, "", true, previousConfigDir, false)
+	test, err = newTransformTest(t, "", true, previousConfigDir)
 	if err != nil {
 		t.Fatalf("failed to start KUBE API Server with encryptionConfig\n %s, error: %v", encryptionConfig, err)
 	}
-	test.cleanUp()
+	defer test.cleanUp()
+
+	// confirm that reading secrets still works after restart
+	// make sure cluster wide secrets read still works
```
Comment (aramase) [inline test/integration/controlplane/transformation/kms_transformation_test.go:526]:
```suggestion
	// confirm that reading cluster wide secrets still works after restart
```

File: test/integration/controlplane/transformation/kms_transformation_test.go [hunk 8]
```diff
@@ -515,35 +516,81 @@ resources:
...
+	}
 }
 
 func TestEncryptionConfigHotReloadFileWatch(t *testing.T) {
 	testCases := []struct {
-		fileUpdateMethod string
-		sleep            time.Duration
+		sleep      time.Duration
+		name       string
+		updateFile func(filePath string, fileContent string) error
```
Comment (aramase) [inline test/integration/controlplane/transformation/kms_transformation_test.go:543]:
```suggestion
		updateFile func(filePath, fileContent string) error
```

File: test/integration/controlplane/transformation/kms_transformation_test.go [hunk 9]
```diff
@@ -515,35 +516,81 @@ resources:
...
-			sleep:            20 * time.Second, // significantly longer than KMSCloseGracePeriod
-		},
-		{
-			fileUpdateMethod: "deleteAndCreate",
+			name: "truncate file",
+			updateFile: func(filePath string, fileContent string) error {
+				// os.WriteFile truncates the file before writing
+				return os.WriteFile(filePath, []byte(fileContent), 0644)
+			},
+			sleep: 20 * time.Second, // significantly longer than KMSCloseGracePeriod
```
Comment (aramase) [inline test/integration/controlplane/transformation/kms_transformation_test.go:551]: Move the comment to the line above

File: test/integration/controlplane/transformation/kms_transformation_test.go [hunk 10]
```diff
@@ -515,35 +516,81 @@ resources:
...
 		},
 		{
-			fileUpdateMethod: "move",
+			name: "delete and create file",
+			updateFile: func(filePath string, fileContent string) error {
```
Comment (aramase) [inline test/integration/controlplane/transformation/kms_transformation_test.go:555]:
```suggestion
			updateFile: func(filePath, fileContent string) error {
```

File: test/integration/controlplane/transformation/kms_transformation_test.go [hunk 11]
```diff
@@ -515,35 +516,81 @@ resources:
...
+					return fmt.Errorf("failed to write encryption config, err: %w", err)
+				}
+
+				return nil
+			},
 		},
 		{
-			fileUpdateMethod: "symLink",
+			name: "move file",
+			updateFile: func(filePath string, fileContent string) error {
```
Comment (aramase) [inline test/integration/controlplane/transformation/kms_transformation_test.go:576]:
```suggestion
			updateFile: func(filePath, fileContent string) error {
```

File: test/integration/controlplane/transformation/kms_transformation_test.go [hunk 12]
```diff
@@ -732,6 +696,46 @@ resources:
 	}
 }
 
+func verifyPrefixOfSecretResource(t *testing.T, wantPrefix string, test *transformTest) {
+	// implementing this brute force approach instead of fancy channel notification to avoid test specific code in prod.
+	// wait for config to be observed
+	verifyIfKMSTransformersSwapped(t, wantPrefix, test)
+
+	// run storage migration
+	// get secrets
```
Comment (aramase) [inline test/integration/controlplane/transformation/kms_transformation_test.go:705]:
```suggestion
```

File: test/integration/controlplane/transformation/kms_transformation_test.go [hunk 13]
```diff
@@ -732,6 +696,46 @@ resources:
...
+	secretsList, err := test.restClient.CoreV1().Secrets("").List(
+		context.TODO(),
+		metav1.ListOptions{},
+	)
+	if err != nil {
+		t.Fatalf("failed to list secrets, err: %v", err)
+	}
+
+	for _, secret := range secretsList.Items {
+		// update secret
```
Comment (aramase) [inline test/integration/controlplane/transformation/kms_transformation_test.go:715]:
```suggestion
```

redundant comment

File: test/integration/controlplane/transformation/kms_transformation_test.go [hunk 14]
```diff
@@ -732,6 +696,46 @@ resources:
...
+			context.TODO(),
+			&secret,
+			metav1.UpdateOptions{},
+		)
+		if err != nil {
+			t.Fatalf("failed to update secret, err: %v", err)
+		}
+	}
+
+	// assert that resources has new prefix
```
Comment (aramase) [inline test/integration/controlplane/transformation/kms_transformation_test.go:726]:
```suggestion
```

this comment should move to https://github.com/kubernetes/kubernetes/pull/113896/files#diff-757c6bfc7bd92dcf0f33365f70feea4a5ceafad120668a60ff2934941ece79efR733

File: test/integration/controlplane/transformation/transformation_test.go [hunk 15]
```diff
@@ -254,43 +256,27 @@ func (e *transformTest) getRawSecretFromETCD() ([]byte, error) {
...
-
-		// create symlink
-		if err := os.Symlink(filepath.Join(parentTempDir, encryptionConfigFileName), filepath.Join(tempDir, encryptionConfigFileName)); err != nil {
-			return tempDir, parentTempDir, fmt.Errorf("failed to create symlink: %v", err)
-		}
+func (e *transformTest) createEncryptionConfig() (
+	filePathForEncryptionConfig string,
+	symLinkPathForEncryptionConfig string,
+	err error,
+) {
```
Comment (aramase) [inline test/integration/controlplane/transformation/transformation_test.go:271]:
```suggestion
func (e *transformTest) createEncryptionConfig() (
	filePathForEncryptionConfig,
	symLinkPathForEncryptionConfig string,
	err error,
) {
```

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go (hunk 6 above)
Comment (nilekhc) [reply staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go]: I intended the same. Fixed the line numbers.

File: test/integration/controlplane/transformation/kms_transformation_test.go [hunk 16]
```diff
@@ -515,35 +516,81 @@ resources:
...
 		t.Fatalf("failed to start KUBE API Server with encryptionConfig\n %s, error: %v", encryptionConfig, err)
 	}
-	test.cleanUp()
+	defer test.cleanUp()
+
+	// confirm that reading cluster wide secrets still works after restart
+	_, err = test.restClient.CoreV1().Secrets("").List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		t.Fatalf("failed to list secrets, err: %v", err)
+	}
```
Comment (aramase) [inline test/integration/controlplane/transformation/kms_transformation_test.go:529]: nit: 

```suggestion
	if _, err = test.restClient.CoreV1().Secrets("").List(context.TODO(), metav1.ListOptions{}); err != nil {
		t.Fatalf("failed to list secrets, err: %v", err)
	}
```

File: test/integration/controlplane/transformation/kms_transformation_test.go [hunk 17]
```diff
@@ -515,35 +516,81 @@ resources:
...
+
+	// make sure cluster wide configmaps read still works
+	_, err = test.restClient.CoreV1().ConfigMaps("").List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		t.Fatalf("failed to list configmaps, err: %v", err)
+	}
```
Comment (aramase) [inline test/integration/controlplane/transformation/kms_transformation_test.go:535]: nit:

```suggestion
	if _, err = test.restClient.CoreV1().ConfigMaps("").List(context.TODO(), metav1.ListOptions{}); err != nil {
		t.Fatalf("failed to list configmaps, err: %v", err)
	}
```

File: test/integration/controlplane/transformation/kms_transformation_test.go [hunk 18]
```diff
@@ -515,35 +516,81 @@ resources:
...
+
+				return nil
+			},
 		},
 		{
-			fileUpdateMethod: "symLink",
+			name: "move file",
+			updateFile: func(filePath, fileContent string) error {
+				// write new config to a temp file
+				if err := os.WriteFile(filePath+".tmp", []byte(fileContent), 0644); err != nil {
```
Comment (aramase) [inline test/integration/controlplane/transformation/kms_transformation_test.go:578]: make `filePath + ".tmp"` a variable and reuse here and https://github.com/kubernetes/kubernetes/pull/113896/files#diff-757c6bfc7bd92dcf0f33365f70feea4a5ceafad120668a60ff2934941ece79efR583

File: test/integration/controlplane/transformation/transformation_test.go [hunk 19]
```diff
@@ -254,43 +256,27 @@ func (e *transformTest) getRawSecretFromETCD() ([]byte, error) {
...
-		// create symlink
-		if err := os.Symlink(filepath.Join(parentTempDir, encryptionConfigFileName), filepath.Join(tempDir, encryptionConfigFileName)); err != nil {
-			return tempDir, parentTempDir, fmt.Errorf("failed to create symlink: %v", err)
-		}
+func (e *transformTest) createEncryptionConfig() (
+	filePathForEncryptionConfig,
+	symLinkPathForEncryptionConfig string,
+	err error,
+) {
+	tempDir := filepath.Join(os.TempDir(), "symlinkdir")
```
Comment (aramase) [inline test/integration/controlplane/transformation/transformation_test.go:272]: Any reason for not using `os.MkdirTemp`?

Comment (nilekhc) [reply test/integration/controlplane/transformation/transformation_test.go:272]: updated

**Comment by aramase** [comment]
/retitle chore: improves tests for hot reload of encryptionconfig

**Review by aramase — COMMENTED** [review]
/lgtm

test failures are unrelated due to python2/python3 migrations (xref: https://github.com/kubernetes/kubernetes/pull/113121#issuecomment-1404338346)

**Comment by nilekhc** [comment]
> /lgtm
> 
> test failures are unrelated due to python2/python3 migrations (xref: [#113121 (comment)](https://github.com/kubernetes/kubernetes/pull/113121#issuecomment-1404338346))

@aramase had to rebase. Need another `lgtm` 😄

**Review by aramase — COMMENTED** [review]
/lgtm

**Comment by nilekhc** [comment]
@enj As we discussed, I have reverted the controller_test changes. PTAL when you get a chance.

**Pull request merged**